pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -745,20 +744,32 @@ func nodeAddresses(srv *servers.Server, interfaces []attachinterfaces.Interface,
 	for _, network := range networks {
 		for _, props := range addresses[network] {
 			var addressType v1.NodeAddressType
-			if props.IPType == "floating" || util.Contains(networkingOpts.PublicNetworkName, network) {
+			if props.IPType == "floating" {
 				addressType = v1.NodeExternalIP
+			} else if util.Contains(networkingOpts.PublicNetworkName, network) {
+				addressType = v1.NodeExternalIP
+				RemoveFromNodeAddresses(&addrs,
```

Comment:
Why the address is removed here if the network name is included in PublicNetworkName?

Reply:
@zuzzas could you please explain your test case?

Reply:
It may be present in the fixed IPs and erroneously counted as an InternalIP. We remove it and add again as a proper ExternalIP address.

Reply:
@zuzzas I'm still confused, in which case `It may be present in the fixed IPs and erroneously counted as an InternalIP`?

Reply:
Without the proposed patch, in the following situation, `10.220.213.38` will be counted both as `InternalIP` and `ExternalIP`. It does not look right, since we've explicitly designated in the CCM's configuration the network with this IP as `public`.

`$ openstack --debug server list f4a319e6-95fb-445a-ac2d-a6374264d6c1`

```json
{
  "server": {
    "addresses": {
      "BGP-VLAN-3894": [
        {
          "OS-EXT-IPS-MAC:mac_addr": "fa:16:3e:42:95:9f",
          "version": 4,
          "addr": "10.220.213.157",
          "OS-EXT-IPS:type": "fixed"
        }
      ],
      "e2e": [
        {
          "OS-EXT-IPS-MAC:mac_addr": "fa:16:3e:6c:f6:bb",
          "version": 4,
          "addr": "192.168.199.52",
          "OS-EXT-IPS:type": "fixed"
        }
      ]
    }
  }
}
```

`$ nova --debug interface-list f4a319e6-95fb-445a-ac2d-a6374264d6c1`

```json
{
  "interfaceAttachments": [
    {
      "port_id": "08368bb9-1021-4f08-a318-44bade127942",
      "port_state": "ACTIVE",
      "tag": null,
      "mac_addr": "fa:16:3e:42:95:9f",
      "fixed_ips": [
        {
          "subnet_id": "728ca9da-da11-4076-ac32-c47a18693c5c",
          "ip_address": "10.220.213.157"
        }
      ],
      "net_id": "a30dbaf7-dd86-4c67-aaa5-42c4fdf33e2a"
    },
    {
      "port_id": "8cc3d233-89ae-4e06-aed8-5b92cea3b5a4",
      "port_state": "ACTIVE",
      "tag": null,
      "mac_addr": "fa:16:3e:6c:f6:bb",
      "fixed_ips": [
        {
          "subnet_id": "b33a5f97-3924-4f8d-ad8c-10d5b08f81aa",
          "ip_address": "192.168.199.52"
        }
      ],
      "net_id": "847533ef-13bd-4655-a9b3-812b5ba1e3e0"
    }
  ]
}
```

Reply:
I'm not sure that should be supported, you are setting a private Neutron network as 'External' for k8s node, it's not common use case AFAIK. Any particular reason you are doing that?

Reply:
I'll re-check this in the morning. Unfortunately, this may not be possible, since we are trying to integrate with a Cloud Provider which uses (opinionated) OpenStack inside.

I'll provide a detailed response soon.

Reply:
@lingxiankong 
What do you mean by "a private Neutron network", how does it map to OpenStack's API representation? This network (`BGP-VLAN-3894`) is *external* to Nodes in our setup. They are connected via VPN to a non-OpenStack managed network gateway, which means that they are External and should be treated as such in the Kubernetes cluster. I'd like to set up a list of **External** networks and make sure that no IP from it is classified as **Internal**. This change aims to fix the current discrepancy.

Reply:
Does `BGP-VLAN-3894` exist in the output of `openstack network list --external` command?

Reply:
```
aks@aks-ThinkPad-X230:~$ openstack network list --external
+--------------------------------------+---------------+--------------------------------------+
| ID                                   | Name          | Subnets                              |
+--------------------------------------+---------------+--------------------------------------+
| a30dbaf7-dd86-4c67-aaa5-42c4fdf33e2a | BGP-VLAN-3894 | 728ca9da-da11-4076-ac32-c47a18693c5c |
| c153f6b2-c186-4828-b98c-9cbd79f13621 | VLAN-3502     | 6dc455a0-f44c-4d9a-9d10-8557c4d38f53 |
+--------------------------------------+---------------+--------------------------------------+```

Reply:
@zuzzas that's pretty weird, according to the console output you provided above,

```
      "BGP-VLAN-3894": [
        {
          "OS-EXT-IPS-MAC:mac_addr": "fa:16:3e:42:95:9f",
          "version": 4,
          "addr": "10.220.213.157",
          "OS-EXT-IPS:type": "fixed"
        }
      ],
```

`BGP-VLAN-3894` should be an internal Neutron network.

Nevermind, it's not harmful to add that logic, do you mind adding some comments for Line 751 before we merge this PR?

Reply:
I'd launch a full-blown investigation into this peculiarity. Alas, it is a public OpenStack from Mail.ru Cloud Solutions.

Sure.

Reply:
It is done.